### PR TITLE
nixos/openrgb: init module

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change

To celebrate 0.7. Also to make a video on how to create a nixos module

closes: https://github.com/NixOS/nixpkgs/pull/125413

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
cc @mweinelt 

I tried to apply https://github.com/NixOS/nixpkgs/pull/125413#discussion_r644489365 , but always got a failure message, even when adding all `hidraw`, `usb`, and `i2c` devices.

I think this is still a vast improvement over what we have now (which is nothing).

**Comment by Kranzes** [comment]
Could we also make an option for the service to load a profile on startup? maybe even generate a profile and apply it?

File: nixos/modules/services/hardware/openrgb.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+let
+  cfg = config.services.hardware.openrgb;
+in {
+  options.services.hardware.openrgb = {
+    enable = mkEnableOption "OpenRGB server";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.openrgb;
+      defaultText = "pkgs.openrgb";
```
Comment (pennae) [inline nixos/modules/services/hardware/openrgb.nix:14]:
```suggestion
      defaultText = literalExpression "pkgs.openrgb";
```

File: nixos/modules/services/hardware/openrgb.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      description = "Set version of openrgb package to use.";
+    };
+
+    motherboard = mkOption {
+      type = types.nullOr (types.enum [ "amd" "intel" ]);
+      default = null;
+      defaultText = "null";
```
Comment (pennae) [inline nixos/modules/services/hardware/openrgb.nix:21]: defaultText for literals isn't needed, and i'd argue it makes things less maintainable. (even that defaultText *always* makes things less maintainable, but sometimes it's necessary)

File: nixos/modules/services/hardware/openrgb.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+    services.udev.packages = [ cfg.package ];
+
+    boot.kernelModules = [ "i2c-dev" ]
+     ++ lib.optionals (cfg.motherboard == "amd") [ "i2c-piix" ]
+     ++ lib.optionals (cfg.motherboard == "intel") [ "i2c-i801" ];
+
+    systemd.services.openrgb = {
+      description = "openrgb server daemon.";
```
Comment (pennae) [inline nixos/modules/services/hardware/openrgb.nix:43]:
```suggestion
      description = "OpenRGB server daemon";
```
is probably better, systemd adds the trailing `.` automatically in many places

**Comment by jonringer** [comment]
> Could we also make an option for the service to load a profile on startup? maybe even generate a profile and apply it?

Looks like it uses some binary encoding for the profiles, so generating a configuration is probably out of the question. But I could probably have it where a profile can be referenced.

**Comment by jonringer** [comment]
thanks @pennae for the review :)

**Comment by aanderse** [comment]
Something something release notes entry and a discourse post... 😄

**Comment by jonringer** [comment]
> Something something release notes entry and a discourse post... smile

Do as I say, not as I do :)

**Comment by jonringer** [comment]
> Looks like it uses some binary encoding for the profiles, so generating a configuration is probably out of the question. But I could probably have it where a profile can be referenced.

Hmm, I'm not sure how to implement this, and we could do a follow up PR to implement it.

**Review by ryantm — APPROVED** [review]

**Comment by Waelwindows** [comment]
Any updates on this? 

It'd better if this was merged already as people are resorting to hacks to get OpenRGB working on NixOS

**Comment by benley** [comment]
It looks like this should have been merged months ago - and the only thing that needs to be done now (afaict) is to move the release notes section to the next upcoming release.

**Comment by jonringer** [comment]
rebased and updated release notes location

**Comment by benley** [comment]
It looks like the docbook build is complaining about a missing reference now:

```

/nix/store/bv07yh3g6xfmsby5vzf4v1kbna21h9l8-nixos-manual-combined/manual-combined.xml:86534: element link: validity error : IDREF attribute linkend references an unknown ID "opt-services-hardware-openrgb-enable"
 86530        <listitem>
 86531          <para>
 86532            <link xlink:href="https://gitlab.com/CalcProgrammer1/OpenRGB/-/tree/master">OpenRGB</link>,
 86533            a FOSS tool for controlling RGB lighting. Available as
 86534            <link linkend="opt-services-hardware-openrgb-enable">services.hardware.openrgb.enable</link>.
 86535          </para>
 86536        </listitem>

/nix/store/bv07yh3g6xfmsby5vzf4v1kbna21h9l8-nixos-manual-combined/manual-combined.xml fails to validate
```

**Comment by jonringer** [comment]
options have been moved to their own separate html page, since I made the original PR :).

**Comment by jonringer** [comment]
10 months sure do go by fast :O

**Pull request merged**